qemu/tests/numa_opts.py

```diff
@@ -1,49 +1,120 @@
+from virttest import utils_package
+from virttest.utils_misc import normalize_data_size
+from virttest.utils_misc import get_mem_info
+from virttest.utils_misc import NumaInfo
 
 
+@error_context.context_aware
 def run(test, params, env):
     """
     Simple test to check if NUMA options are being parsed properly
+    1) Boot vm with different numa nodes
+    2) With qemu monitor, check if size and cpus for every node match with cli
+    3) In guest os, check if size and cpus for every node match with cli
 
-    This _does not_ test if NUMA information is being properly exposed to the
-    guest.
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
     """
 
-    dbg("starting numa_opts test...")
+    def numa_info_guest():
+        """
+        The numa info in guest os, linux only
+
+        return: An array of (ram, cpus) tuples, where ram is the RAM size in
+                MB and cpus is a set of CPU numbers
+        """
+
+        numa_info_guest = NumaInfo(session=session)
+
+        numa_guest = []
+        nodes_guest = numa_info_guest.online_nodes
+        for node in nodes_guest:
+            node_size = numa_info_guest.online_nodes_meminfo[node]['MemTotal']
+            node_size = float(normalize_data_size('%s KB' % node_size))
+            node_cpus = numa_info_guest.online_nodes_cpus[node]
+            node_cpus = set([int(v) for v in node_cpus.split()])
+            numa_guest.append((node_size, node_cpus))
+
+        # It is a known WONTFIX issue for x86, node info of node0 and node1 is
+        # opposite in guest os when vm have 2 nodes
+        if (vm_arch in ("x86_64", "i686") and len(numa_guest) == 2):
+            numa_guest.reverse()
+        return numa_guest
 
     vm = env.get_vm(params["main_vm"])
+    os_type = params["os_type"]
+    vm_arch = params["vm_arch_name"]
+    session = vm.wait_for_login()
 
+    error_context.context("starting numa_opts test...", logging.info)
+
+    # Get numa info from monitor with "info monitor"
     numa = vm.monitors[0].info_numa()
-    dbg("info numa reply: %r", numa)
+    error_context.context("info numa reply: %r" % numa, logging.info)
+
+    if os_type == 'linux':
+        if not utils_package.package_install('numactl', session):
+            test.cancel("Please install numactl to proceed")
+        # Get numa info in guest os with "numactl -H",  only for Linux
+        numa_guest = numa_info_guest()
+        error_context.context("numactl -H reply: %r" % numa_guest, logging.info)
+
+        # Use 30 plus the gap of 'MemTotal' in OS and '-m' in cli as threshold
+        MemTotal = get_mem_info(session, 'MemTotal')
+        MemTotal = float(normalize_data_size('%s KB' % MemTotal))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % MemTotal, logging.info)
+        threshold = float(params.get_numeric("mem") - MemTotal) + 30
```

Comment:
Why need add 30?

Reply:
Could you check the comment i added in the bz before.  Without plus 30, the threshold is not large enough while use `numactl -H`. 
I just noticed, `MemTotal` in `/sys/devices/system/node/node%s/meminfo` have a small gap with the size in `numactl -H` output.   let me check if we still need the `30` now.

Reply:
It still need plus 30,  otherwise the threshold is not large enough, thanks!:
```
numa info in guest: [(3725.2265625, {0, 1, 2, 3, 4, 5})]

05:51:57 INFO | Context: MemTotal in guest os is 3735.2265625 MB
05:51:57 INFO | Context: The acceptable threshold is: 360.7734375

TestFail: [Guest]Wrong size of numa node 0: 3725. Expected: 4096
```

Reply:
Okay.